### PR TITLE
amp-instagram: Resizing in response to messages posted by iframe

### PR DESCRIPTION
**Pull request by src-code**
Enabling the automatic resize of the `amp-instagram` component in response to messages received from the Instagram iframe.  Instagram posts messages with the following data property:

```json
"data": {
  "type": "MEASURE",
  "details": {
    "height": 123
  }
}
```

Fixes issue #5032.

### CODE REVIEW

File: extensions/amp-instagram/0.1/amp-instagram.js [hunk 1]
```diff
@@ -141,6 +151,30 @@ class AmpInstagram extends AMP.BaseElement {
...
+    const data = isObject(event.data) ? event.data : tryParseJson(event.data);
+    if (data === undefined) {
+      return; // We only process valid JSON.
+    }
+    if (data.type == 'MEASURE') {
+      const height = data.details.height;
+      if (this.iframe_./*OK*/offsetHeight !== height) {
+        // Height returned by Instagram includes header, so
+        // subtract 48px top padding
+        this.attemptChangeHeight(height - PADDING_TOP).catch(() => {});
```
Comment (cramforce) [inline extensions/amp-instagram/0.1/amp-instagram.js]: Probably OK to call `changeHeight`

File: extensions/amp-instagram/0.1/amp-instagram.js [hunk 2]
```diff
@@ -120,6 +126,10 @@ class AmpInstagram extends AMP.BaseElement {
   layoutCallback() {
     const iframe = this.element.ownerDocument.createElement('iframe');
     this.iframe_ = iframe;
+
+    this.win.addEventListener(
+        'message', event => this.handleInstagramMessages_(event));
```
Comment (aghassemi) [inline extensions/amp-instagram/0.1/amp-instagram.js]: Let's keep a reference to the `unlisten` function so we remove the listener in `unlayout` e.g.
(`listen` is a utility function that can be imported from `../../../src/event-helper`)
```
this.unlistenMessage_ = listen(this.win, 'message', this.handleInstagramMessages_.bind(this));
```
then in `unlayout`
```
if (this.unlistenMessage_) {
this.unlistenMessage_();
}
```

File: extensions/amp-instagram/0.1/amp-instagram.js [hunk 3]
```diff
@@ -141,6 +151,30 @@ class AmpInstagram extends AMP.BaseElement {
...
+        !(isObject(event.data) || event.data.indexOf('{') == 0)) {
+      return;  // Doesn't look like JSON.
+    }
+    const data = isObject(event.data) ? event.data : tryParseJson(event.data);
+    if (data === undefined) {
+      return; // We only process valid JSON.
+    }
+    if (data.type == 'MEASURE') {
+      const height = data.details.height;
+      if (this.iframe_./*OK*/offsetHeight !== height) {
```
Comment (jridgewell) [inline extensions/amp-instagram/0.1/amp-instagram.js]: Does this need to be in vsync?

Comment (aghassemi) [inline extensions/amp-instagram/0.1/amp-instagram.js]: Makes sense to be in vsync. `attemptChangeHeight` does its own mutate so I think we can do the check and the `attemptChangeHeight` in the same measure block. e.g.: 
```
    const height = data.details.height;
    this.getVsync().measure(() => {
      if (this.iframe_./*OK*/offsetHeight !== height) {
         this.attemptChangeHeight(height - PADDING_TOP).catch(() => {});
      }
    });
```

File: extensions/amp-instagram/0.1/amp-instagram.js (hunk 2 above)
Comment (src-code) [inline extensions/amp-instagram/0.1/amp-instagram.js]: Sounds good, I'll add that!

File: extensions/amp-instagram/0.1/amp-instagram.js [hunk 4]
```diff
@@ -55,6 +62,9 @@ class AmpInstagram extends AMP.BaseElement {
 
     /** @private {?string} */
     this.shortcode_ = '';
+
+    /** @private {?function} */
```
Comment (aghassemi) [inline extensions/amp-instagram/0.1/amp-instagram.js]: `{?Function}`

**Pull request merged**